<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Allison C Fialkowski" />

<meta name="date" content="2017-11-30" />

<title>Variable Types</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Variable Types</h1>
<h4 class="author"><em>Allison C Fialkowski</em></h4>
<h4 class="date"><em>2017-11-30</em></h4>



<p>All variables are generated via an appropriate transformation of standard normal variables, as described below. See the documentation for <code>corrvar</code> and <code>corrvar2</code> for more details about all inputs. The parameter inputs should be checked first using <code>validpar</code>. Some code has been modified from the <strong>SimMultiCorrData</strong> package <span class="citation">(Fialkowski 2017)</span>.</p>
<p><strong>1) Continuous Variables:</strong> Continuous variables are simulated using either <span class="citation">Fleishman (1978)</span>’s third-order (<code>method</code> = “Fleishman”) or <span class="citation">Headrick (2002)</span>’s fifth-order (<code>method</code> = “Polynomial”) power method transformation (PMT). This is a computationally efficient algorithm that simulates continuous distributions through the method of moments. It works by matching standardized cumulants – the first four (mean, variance, skew, and standardized kurtosis) for Fleishman’s third-order method, or the first six (mean, variance, skew, standardized kurtosis, and standardized fifth and sixth cumulants) for Headrick’s fifth-order method. The transformation is expressed as follows: <span class="math display">\[Y = c_{0} + c_{1} * Z + c_{2} * Z^2 + c_{3} * Z^3 + c_{4} * Z^4 + c_{5} * Z^5,\ Z \sim iid\ N(0,1),\]</span> where <span class="math inline">\(c_{4}\)</span> and <span class="math inline">\(c_{5}\)</span> both equal <span class="math inline">\(0\)</span> for Fleishman’s method. The real constants are calculated by <code>SimMultiCorrData::find_constants</code>, which solves the system of equations given in <code>SimMultiCorrData::fleish</code> or <code>SimMultiCorrData::poly</code>. All variables are simulated with mean <span class="math inline">\(0\)</span> and variance <span class="math inline">\(1\)</span>, and then transformed to the specified mean and variance at the end.</p>
<p><strong>Continuous Mixture Variables:</strong> Mixture distributions describe random variables that are drawn from more than one component distribution. For a random variable <span class="math inline">\(Y\)</span> from a finite continuous mixture distribution with <span class="math inline">\(k\)</span> components, the probability density function (PDF) can be described by: <span class="math display">\[h_{Y}(y) = \sum_{i=1}^{k} \pi_{i} f_{Y_{i}}(y),\ \ \sum_{i=1}^{k} \pi_{i} = 1.\]</span><br />
The <span class="math inline">\(\pi_{i}\)</span> are mixing parameters which determine the weight of each component distribution <span class="math inline">\(f_{Y_{i}}(y)\)</span> in the overall probability distribution. As long as each component has a valid PDF, the overall distribution <span class="math inline">\(h_{Y}(y)\)</span> has a valid PDF. The main assumption is statistical independence between the process of randomly selecting the component distribution and the distributions themselves: Assume there is a random selection process that first generates the numbers <span class="math inline">\(1,\ ...,\ k\)</span> with probabilities <span class="math inline">\(\pi_{1},\ ...,\ \pi_{k}\)</span>. After selecting number <span class="math inline">\(i\)</span>, where <span class="math inline">\(1 \leq i \leq k\)</span>, a random variable <span class="math inline">\(y\)</span> is drawn from component distribution <span class="math inline">\(f_{Y_{i}}(y)\)</span>.</p>
<p>Mixture distributions provide a useful way for describing heterogeneity in a population, especially when an outcome is a composite response from multiple sources. They may also be used to model distributions with outliers. For example, the <em>contaminated normal outlier distribution</em> is used when a proportion of measurement errors, which are usually modeled as normal variables with zero mean and common variance <span class="math inline">\(\sigma^2\)</span>, have larger variability. These distributions can be thought of as mixture distributions with contamination percentage <span class="math inline">\(\pi\ \epsilon\ (0,\ 1)\)</span>. For example, the PDF of <span class="math inline">\(Y\)</span> may be given by: <span class="math display">\[h_{Y}(y) = \pi \phi(0,\ 9\sigma^2)(y) + (1-\pi) \phi(0,\ \sigma^2)(y),\ \ -\infty &lt; y &lt; \infty.\]</span> Here, <span class="math inline">\(\phi(\mu,\ \sigma^2)(y)\)</span> is the PDF of the normal distribution. Setting <span class="math inline">\(\pi = 0.3\)</span> and <span class="math inline">\(\sigma^2 = 1\)</span> produces the following distribution for <span class="math inline">\(n = 10000\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;SimCorrMix&quot;</span>)
Nstcum &lt;-<span class="st"> </span><span class="kw">calc_mixmoments</span>(<span class="dt">mix_pis =</span> <span class="kw">c</span>(<span class="fl">0.3</span>, <span class="fl">0.7</span>), <span class="dt">mix_mus =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), 
  <span class="dt">mix_sigmas =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">1</span>), <span class="dt">mix_skews =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">mix_skurts =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), 
  <span class="dt">mix_fifths =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">mix_sixths =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))
Nmix &lt;-<span class="st"> </span><span class="kw">contmixvar1</span>(<span class="dt">n =</span> <span class="dv">10000</span>, <span class="st">&quot;Polynomial&quot;</span>, Nstcum[<span class="dv">1</span>], Nstcum[<span class="dv">2</span>]<span class="op">^</span><span class="dv">2</span>, 
  <span class="dt">mix_pis =</span> <span class="kw">c</span>(<span class="fl">0.3</span>, <span class="fl">0.7</span>), <span class="dt">mix_mus =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">mix_sigmas =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">1</span>), 
  <span class="dt">mix_skews =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">mix_skurts =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">mix_fifths =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), 
  <span class="dt">mix_sixths =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</code></pre></div>
<pre><code>## 
##  Constants: Component  1  
## 
##  Constants: Component  2  
## 
## Constants calculation time: 0 minutes 
## Total Simulation time: 0.001 minutes</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_simpdf_theory</span>(<span class="dt">sim_y =</span> Nmix<span class="op">$</span>Y_mix[, <span class="dv">1</span>], <span class="dt">overlay =</span> <span class="ot">FALSE</span>, 
  <span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PDF of Y&quot;</span>)
<span class="kw">qqnorm</span>(Nmix<span class="op">$</span>Y_mix[, <span class="dv">1</span>], <span class="dt">main =</span> <span class="st">&quot;Q-Q plot for Y&quot;</span>)
<span class="kw">qqline</span>(Nmix<span class="op">$</span>Y_mix[, <span class="dv">1</span>], <span class="dt">col =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAATgAAAEgCAMAAAAXJJ+uAAAA1VBMVEUAAAAAACsAAFUAAIsAKysAK4AAVaorAAArACsrAFUrK1UrVYArVaorgNQzMzNNTU1NTWtNTYhNa6ZNiMRVAABVACtVKwBVKytVgIBVgKpVqtRVqv9rTU1rTWtrTYhra01ra6ZriIhrpuGAKwCA1P+ITU2ITWuITYiIxP+ma02ma2um4f+qVQCqVSuqgCuq1P+q/9Sq///EiE3E///UgCvUqoDU1P/U/6rU///hpmvhxMTh///r6+v/qlX/xIj/1ID/1Kr/4ab//6r//8T//9T//+H///+rRuuqAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAJXklEQVR4nO2dDXvbNBSFXUzT8VEytrANusGgbAxaGNsyWOnXmnb6/z8Jy4pjyZFs6ca2rpJzHtgSHdmW311LsiLLmYBIymIXIFUBHFEARxTAEQVwRAEcUQBHFMARxQjcaab05a/1l72vXktrvrSyz97Ytvz4uMh5JD/dHWfZoVhusv9hwNIyBJdlR9qXPYmxHZykpfIJcX2gPi1m5W6GE0twRajUXySrdnAFpDq4ig0notxiMmhpWYErY+V9GTP1F3npzVvDp4iyyepLQbHYsvjTflX3Jn7g1N/VlwJKEU02cO8fZdnn33+oInVSpV/Kz6dVVTeY+IEzIk7WX0XoWMD92biqJ5Uha7wnB8O2DIIZOKOOW9b25YeqjlumiTKu9l/f/VUGlnGpll/1nAOJIzh50l3glK/isQGu3JGRMIT4gVNdtw5wVUM6l0lNcMO3DIIZOO366qjjanCFYQE3dA3HH5yjVe2KuJ0H9/GRvR9XBKK7jttxcI07h0Z3pKVVBTjjXtXdjwM4C7i9r/+R36x3Do+LO4cf5KddB5eWAI4ogCMK4IgCOKIAjiiAIwrgiKKC+2ZntSk4W+KFK/e2GIUDcAA3ogFwRAPgiAbAEQ2AIxoARzRSBpfnwx8D4Po2EgaXXwgHOYBrNQCOZuQC4EiGBOcgB3BtBsDRjFwAHMlICNyn36ffvlMfr6bT+ydVOsA58yi9fSj/k7p5diKuKohRwdnJ8QJ3+/JE3Pz4Tv+qBHDOPKUktJqWOHtwvvwEcM48pQxwN09VHVf+2HMRQ7nxVxTRIi5yHZeLNCMueh2XEDhejUNC4LTuiLxMb55HbRwqcFZyzMDdvig7wG9/KZrU6aovDHBJ3XLlKwPgAI5YEoAjlgTgiCUBOGJJQoy8NgAO4IglAThiSQCOWBKAI5aECM5GDuAcRq4bAOdvABzRADiiAXBEA+CIBsARDYAjGgBHNAxwFnIA5zAAjmgAHNHYFnCjz3bJW76NqOQiLjeNdCOun5IAHLEkAEcsCRncOjmAsxsARzQAjmgAHNEAOKIBcEQD4GhG3jQAzs9YA7dGDuCsBmtwISs2A5yRPs98XwMGcE3Lkx3ArZter3wBuIYzL99gd9lJDuD0ZPlipkn5ad71thyA01IXs9ULFQDO5aAftxvgcovRIBfzUv1OXamd16kAOIALMmzgytfiLuXxhk2A01KriNOlLRF083Q6fVWlA5wzj1L9TP7ti1fi6n6kxQySA6etAnH14LyIvyrkAK5KWsz2/52tveK6ue7ICwkuwko31jk2USbeJLa2ki3iGiHH61I1wZ1FWz4jbXBn9bp7I4PLrQYPcItZ3Y+r6zh9iSAt3gAucMXCWgDnzKNULxH0dioVpzuSIDiXOIAzyfHqxzkFcM487QI4Z552AZyRXvZJvH6RBjg9+bIciZvvrY8urQngtNS740n59ym7X/KZg1vM1NAvv6Fz5uCSiziDXMw67vogrTqOBTj7Tb5To4Jz8uEALlAA58zTLoDTk6vLFZeq27GCK5rT+URc3+PWOHAHJ/txl0VXpIDXKYDTUiW46y/elP93CeC01Lvjw3IaBMC5DFcdJ++1Tg/FnNmdg/vOigs42ToULatHowpwKfTjAI5otAyCuG5iByqVExzPEWD+4EJGgEecINQyK2n8CUspjcexjziuI8DswTGNuLaZDjzAMR0BZg6O7wgwc3CBAjhnnnYxAWef4zpYqZzg5AOrmU8VB3C2DnB25NxwJYDTUnl2R9qfA2EBjmcHOAFwiLguI6k6LgVwLFvVJMD5C+C01Ltjj2t0KS7gLMtDDFiqjlbVRwCnpcrfVX0FcHqy16wRpdHAda3LwgHcguOTNSmACxHAOfO0C+DqRLV0nKfYgFtliAZO/tSgfnTwEcBVSaov4jNRqdRY4Jrc+IFTvd/LZoOqLREkVk/nC4BbA9e8w6+fyZePmY+/fEaq4IxVIKY/I+J8wenrjvx3vkQ45hJB3bNqRp53QwCHOs4RcbZf8mODW+PGD5xdCYCr8vACpzcOABdyr6p3RwAuAFy9RJAAuBBwLgGcM0+7xgG3zg3gvAwvcMtcAKcZAEc0AI5oABzRADii4QdOZQM4zQA4mmHhBnA+BsAB3EqswJUZAa42AI5m2LgBnIcBcABXC+Ccedo1AjgrN4DrNgLAybwAVxkARzPs3ACu09hWcINPCwqbhTTanKUti7gidyoR109J3IaDG8B1GYHgRA5wygA4muHi1ucW2wguBziaQbjwAE7Q2khnkAJcx64ATgYPqcscMixAMtiDo46nWNHtDrhNhiU9J00QDd7g8o3Gc31mDZMN5uA221X3Q0x0gzW4jWe7dT5TTTeYgstLbb6rRhOx9eCM891oV3l/u1pz+IHrN0w0dFsOru/rq97fdoNrdl03P9vdALfW5e/hbKu2ZlvBaU3ppruyiHjb63bigyt5XbgGg/o7W/s/DO0Y0cEtz8V1RiG78jOsRxoUnLbSjbHoDRmcGQA982kx8vXIGxSc9mi58ZQ5EdzmZd+0Z5xbDc9d+YPTFjMw1zUggVu/YEYGJ4yIHxKctnyGuZKGJ7jclH8RBzWMIlkKNxS4xhJB+Q4oeNLNphG3ZcZ4l+qWGbEah+SNeN2RxI0AcNpKN8uPSgDnzNMugHPmaRfAOfO0C+Ccedr1zc5qQ3B2mju0BcARtwA44ha9gtslARxRAEcUwBEFcET1CK4cczJ+Cuve5Ol0+uDc/xBhe6ccwfss+gOn3vZgjD116so/a6mwvVOO4H0WvYFTb3swRzs7pdZl9lbg3glH8D+L3sCptz2Y4+td+vSHb06lsL1TjuB/Fj3XcWGndvvyp+n9gDMLBxd6BP+ziAru5tmJuP0toC0JBhd6hHHBnU2nDwMOaWwUBCIcXOgRokVcePUdclqExiHwCP5n0Xc/LqjDcFU0/TfPA3pZwd2R4CN4n0Xf4IyfwjpVXK9BHdqwvVOO4H0WuOUiCuCIAjiiAI4ogCMK4IgCOKIAjiiAIwrgiAI4ohIDNy9fnLuYTWIXJDVw1wfWV4RHUGLgxOn+h+UfkZUauMvsqLhSj2IXIz1wd8cTMa/fcR1PqYErqP19fBi7ECJBcNcHT+7FbxoSBCdOMwZNQ4rgZPPAQAmC49A0JAjujkXTkBy4xSybxC6DUmLg+AjgiAI4ogCOKIAjCuCIAjii/gcpENUmfZedfQAAAABJRU5ErkJggg==" alt="Contaminated normal distribution Y" /><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAATgAAAEgCAMAAAAXJJ+uAAAAq1BMVEUAAAAAACsAAFUAKysAK1UAK4AAVaorAAArACsrAFUrKysrK4ArVYArVaorgNRVAABVACtVKwBVKytVVStVgFVVgIBVgKpVgNRVqqpVqtRVqv+AKwCAVQCAVSuAgFWAqoCA1P+qVQCqVSuqVYCq1ICq1Kqq1P+q/6qq/9Sq///UgCvUqlXUqoDU1IDU1NTU/6rU/9TU////AAD/qlX/1ID/1Kr//6r//9T///9cAhetAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMP0lEQVR4nO2dC3vkphWGWTczk6SNnU2bNLXbxrNt065yq1eurf//yyoOF11GQvCBJBifL4+9CkIgvT7AQSAQDQuS2PsGShWDA8XgQDE4UAwOFIMDxeBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UAwOFIMDxeBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UAwOFIMDxeBAMThQe4B7/fkbIcSXf3laDLQnH8RhdOL1Hx/N0YeTEF/NZXYW4tYc3Mfc9Ug7gHt5L5Q++3Eh0OoS3P/e3xhwtbzsdja3O/Husf33+XTBPkrbg2shGNlnnw4cXDN66nMXr1Jk5lQJcaQUnLGCtT249kHeyQL586lnJ1OBLZt/faNM0ICTxfl33z1RuTOI6VBCsSclyh/eixtlu/Jvci/NctYoIW0OrvvTt4XImNFk4FmbYHtGg6tUQHs4Aa47qQ9NOrKQ/vdh0pIjtDm4lsxRH55t6ZkMbIm0RvOTLGkKnKqmKrKdi6LaP9n+PvzY/NrYCOLLtC1DswO49gFNmekqp8lA3QxKRAqchkX/XIDrn6yGmKgCPTZptTe41tZa8/hlGhyhkAEEzlR0FGMMbnBy3FzUInHL0OxXVKuvJIibjwrcb8NAFVOzaZ/6nrCY8iwDLsANTo7BtZenNri9GgfpXf1RKBqC6u5+oIqZzuKuApzyPFoHduC2TgXG1HHXCI5srHMd5gOdrWqPzVSreo3gtF2R9/XdkyPQ+HGtCV34cdTP0janMQ38uKsEZ/rzf+p3HSYC21L3n78J8Ydxz+HPFP2D7dYaTN3JawVn9Prh95fd7i7wnNrZT6qM38cxOFAMDhSDu0oxOFAMDhSDA8XgQDE4UAwOFIMDxeBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UAwOFIMDxeBAdeBe7m5pKmPKqdlXrA7c+fD0fDo25+STBa5TFtzLnZya3f7kPJiZkQbgznJuGYPzUq+oHl/u5AxJLqpe6jcONGeUufkpsTsiitde4NImt50MMgQcfQ10jvviqVRwQv8g4Op3j5VsHKLIFQqus7ZwcK8PbcdBztuOckfeIDjpx0lwcQ7wGwRnLO4c1VktEBw1C/F1XBX3eWJ54DSyyFY1+iu74sCNcbEf5ykGh0mXUAYXKKrcBncdCK738R70cUF4Vy8LqQZ1cNNsccsyfsjgnhncolpsihyDC5CuVzS8/gnvFOh327W31dxb6HLZGvmiVmaLc8lCu2zNgE7+16rP8AY6+Q5uEeCu/7WS80V5KLhzl9a1v8h0DzDgFhd5VwnSWFcLAzPcOMzI1m5z570Tskdm0cBrruM6JyQhuHOKuQ95g/MYPkXGHFIshZgzOK9hZ2ywJl4Zg/PiBg3WpJg1ki044ccNGqxJsYpkpuB8sWFFVVxtq+rPjf24nkQANwZnZXj5cUPAmbI6X1SVx1K74mQHLsjcGswBPjxVx+b5NN9EEDg5TWLed8kMXCi25tMn76TNgaRSy9lK83NHZBSNbC5WXuDCuH1qBTrAz198pJ8ZSXDaIkevO4P+qpspABtBo2u8EzcHcraSfLW0AK4ci/PGZqHRVd7J2yP56vd86y6q8h6OjaNjmw84P24DaHSZd/rd4fko0bj9X7UfRz37mjgbcB7YLqDRdd4ZoHe2SXKoFs1tEhpd6Z0FeGvbJAfKjW0WGl3qnYc5uJ6+qsvcnNDoYu9cRv8f+fFgBuBmuS1Co6u9sxkHxH2vuj+4aWxe0Ohy73zGAWWP5E+Zmzc0SgCOWPJIvrjgFgSNkkAjRn6vuie4MbZgaJRIcETTqsYtZrAbuCE1AUGjdJJH3CU531zTQKOkkkfcJTmvLMfUol7RIG9HZL7y4/KYgemNwV2a2oTTG5ZiaMRK9durcnoO09Bi3wiGgqvNN1xV3O4uW4GbgRb/IjUQnB3G7202uW6+EVnMQkvx/jkQnB18qQ4/uYrq4lSwlcG5oKXJGgXn7nJV5gXm7JvMFcE5oSXLFy2qzk5+r8HdeMzBDS1lpnDj8Pz542zcebtc5Rku0l4bGuUWGtG6I44pS9tb3LbQKMfgiKredzepFusmdZwb2lqVQvKIpMU3Aekq6Y0tzeabPOKWye0EjbJOHnGz5PaDRrknj7hRcntCo/yTR9wkuX2h0R0kj7hBcntDo3sAIu69fpx+C7kjtQYCt/P6cVs5agtCHOD91o+7hLbfoE9wxL3Wj1NDK3lQa2Isbsv147oJpJlgi6jjNls/rjeIlw82uFXdaP244chnRthy9uMuZ93mxC1TcFNj7FlhyxHczMSEzLiFD9Z097+COzI/VTk3bhlZnGsKTH7cMgG3MG8oP2wR7kjkXj9dcsvzu3PkBjrA7e8KcuRGCHxmqOXJDepyqS5DbJfLb1pfTp2FgYBOfpL14/zmQopsuUVYnOPrwUT5ZowNq+Ooe796J19kzS1i1nmcD7yYb+bcMvHjJs7njS1XcLmbWwOBW153JDLfArBhrWqKDUId+ZaADR2sWTPfEqg16GDNevlqYNlzQ+o41yTW2HwLMbcGA3dabhzA6frlcIOK6vInNeB0/V4pzZ3bOo0DNnla126uxXcz0iqNg8dnJJf5moWei8C2UuMAWFwpjalVRCff+UlS4HR9XUQLKaZSK/VVA6fr601hVB1XhvLo5AvzXyn2lksnnwytnPpNCgC3vLhoeL7G5MoR0jgsLi7qSGWuY1CWuTWoA7y4uGgHyOnHaViFMSNhDvDC4qJtnKUKUJhfwv5TloA6bnlxUY8ObQcrdGvhTIS4Iz6Liy7tXfAmwaXKl8Gh+b6pOu75dEsL53tMVnK9R3lzrSr5vbIC83CAl8EVrFBwtNoI7cy1vLioE1zxCgNH3wwqX2N5mtfiK88wsys0tgF3bzcIYXABkQmcqt7i5seF30CpsXt1HE3/TbGvWTYPtwG4+t3jy520tRRbN2bzcBuAky5cy+35tPLMm+uJvYbjlc3DMbgMYzM4MHbxfaS9xOBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UAwOFIMDtRI4/7Uk6GMdz8/I/AbMgYSVghbAWAdc7b3wy+tDi8Jzoxy5MkrtSy4kYZ180Go1q4AL2PlGD635jHSoYUnf3TgDElYK3K5nFXDV4a9hYxdedhSMwjdhpcCbXgPc8+ePgevleI2tqU9/whaU8R+0C73pFcDJIhUGbn4X9EEssp4AG/JNWCr4plcAF7wobu3fNoSB80xYKvim04OjEhVyD75mEVxU/e0t/KbTgquk51Tp+VKL8+wq5WdVvo8X2jh4J6zvxeumrXZ3gP0XcwpzR4BVovau44LugabQeirIAQ5JWKsocLqM+OGoArpcQQkr5QDu+sXgQDE4UAwOFIMDxeBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UAwOFIMDxeBAMThQDA4UgwPF4EAxOFAMDhSDA8XgQDE4UNuA6xZS/7fv1jj1/fROCV1gvTw1RCeSZj+eobazODUv0PcZZuOZE2qFWPecJBOXwfVO6AmGzoUCrwvctyeht4fUH77U6uDl67/LJWFVsNxA5/DbnYlHywKqCZMaglqzrdFryZpwG+f7tm64t4nQOZMf7c0TD3J7cPKDF/2jF+aUTyYfX05PM8G2apIBkpJZb0yDs9t21DcfLbgujsrAJCJ/TMJ2UctIbQ/ulsqXXeVPTVCl1cRue4v/dc+sLIv2zZUPrcHZmZm14iPDe3FMJhacTThu491Ou9Rx7S/14O3DqUpKP2Vjg80z9+uwWggnuGEcomXB9RJOQ243cGaatyKjn7KxwRacKZJtFXXzz87iJovqIM4InJ1WTl8VlljH9S2uMc3i0OJ0vL7F0b+9oto1Docnk+YwzpTFGZ2hvdkH2guc9RB6ddx9z3EY13FUNdVdUSV35PXh8As1JKpKG8YZght6JAn8k73AqRU55V++a1XJ+dDBEoZtVV8fjsqQxG33zGciJz+VlDuctkf3wzi6Rei3qjJhMr2gD8KmtRs48quonjZ+nCJigs8jP46m6p+1YZJ0tdVGlj2671Ud18Uh/2TkxymTDJvEP6PyO/kf4iEgKh/cTmJwoBgcKAYHisGBYnCgGBwoBgeKwYFicKAYHCgGB4rBgWJwoBgcKAYHisGBYnCgGBwoBgeKwYFicKAYHKj/A2Lrcn7AHfucAAAAAElFTkSuQmCC" alt="Contaminated normal distribution Y" /></p>
<p>The Q-Q plot shows that the assumption of homogeneous variance for the error distribution would not be valid in this situation <span class="citation">(Davenport, Bezder, and Hathaway 1988; Pearson 2011; Shork, Allison, and Thiel 1996)</span>.</p>
<p>Continuous mixture variables are generated at the component level. Each component variable is created using the PMT, and the mixture variable is generated from these based on a random multinomial variable described by the mixing probabilities. Correlations are also controlled at the component level. Users specify the target pairwise correlations between the components across continuous mixture variables and between components and other types of variables. Since components within a mixture variable should be uncorrelated, those pairwise correlations should be set to <span class="math inline">\(0\)</span>.</p>
<p>The function <code>rho_M1M2</code> calculates the expected correlation between two mixture variables <span class="math inline">\(M1\)</span> and <span class="math inline">\(M2\)</span> based on the component distributions and the correlations between components. The function <code>rho_M1Y</code> calculates the expected correlation between a mixture variable <span class="math inline">\(M1\)</span> and another random variable <span class="math inline">\(Y\)</span>, that may have an ordinal, continuous, or count distribution. If the user desires a specific correlation between a mixture variable <span class="math inline">\(M1\)</span> and another simulated variable <span class="math inline">\(M2\)</span> or <span class="math inline">\(Y\)</span>, various component level correlations can be tried in these 2 functions to see if the desired correlation can be achieved. See the <a href="cont_mixture.html">Expected Cumulants and Correlations for Continuous Mixture Variables</a> vignette for more details on continuous mixture variables.</p>
<ol style="list-style-type: lower-alpha">
<li><p>The <strong>required parameters</strong> for simulating continuous variables include: skewness, standardized kurtosis (kurtosis - 3), and standardized fifth and sixth cumulants (for the fifth-order method). If the goal is to simulate a theoretical distribution (i.e. Gamma, Beta, Logistic, etc.), these values can be obtained using <code>SimMultiCorrData::calc_theory</code>. If the goal is to mimic an empirical data set, these values can be found using <code>SimMultiCorrData::calc_moments</code> (using the method of moments) or <code>SimMultiCorrData::calc_fisherk</code> (using Fisher’s k-statistics). If the standardized cumulants are obtained from <code>calc_theory</code>, the user may need to use rounded values as inputs (i.e. <code>skews = round(skews, 8)</code>). For example, in order to ensure that skew is exactly <span class="math inline">\(0\)</span> for symmetric distributions. Due to the nature of the integration involved in <code>calc_theory</code>, the results are approximations. Greater accuracy can be achieved by increasing the number of subdivisions (<code>sub</code>) used in the integration process.</p></li>
<li><p>For mixture variables, the parameters are specified at the component level by the inputs <code>mix_skews</code>, <code>mix_skurts</code>, <code>mix_fifths</code>, <code>mix_sixths</code>, and <code>mix_Six</code>. The mixing probabilities, means and standard deviations of the component variables are given by <code>mix_pis</code>, <code>mix_mus</code> and <code>mix_sigmas</code>.</p></li>
<li><p>The means and variances of all continuous variables are specified by <code>means</code> and <code>vars</code>. These are at the variable level, i.e., they refer to the continous non-mixture and mixture variables themselves. The function <code>calc_mixmoments</code> calculates the expected mean, standard deviation, and standardized cumulants for mixture variables based on the component distributions.</p></li>
<li><p>For some sets of cumulants, it is either not possible to find power method constants or the calculated constants do not generate valid power method PDFs. In these situations, adding a value to the sixth cumulant may provide solutions (see <code>find_constants</code>). When using Headrick’s fifth-order approximation, if simulation results indicate that a continuous variable does not generate a valid PDF, the user can try <code>find_constants</code> with various sixth cumulant correction vectors to determine if a valid PDF can be found. These sixth cumulant corrections are specified in the simulation functions using <code>Six</code> or <code>mix_Six</code>.</p></li>
<li><p><strong>Choice of Fleishman’s or Headrick’s Method:</strong> Using the fifth-order approximation allows additional control over the fifth and sixth moments of the generated distribution, improving accuracy. In addition, the range of feasible standardized kurtosis values, given skew and standardized fifth (<span class="math inline">\(\gamma_{3}\)</span>) and sixth (<span class="math inline">\(\gamma_{4}\)</span>) cumulants, is larger than with Fleishman’s method. For example, the Fleishman method can not be used to generate a non-normal distribution with a ratio of <span class="math inline">\(\gamma_{3}^2/\gamma_{4} &gt; 9/14\)</span> <span class="citation">(Headrick and Kowalchuk 2007)</span>. This eliminates the Chi-squared family of distributions, which has a constant ratio of <span class="math inline">\(\gamma_{3}^2/\gamma_{4} = 2/3\)</span>. The fifth-order method also generates more distributions with valid PDF’s, However, if the fifth and sixth cumulants do not exist, the Fleishman approximation should be used.</p></li>
</ol>
<p><strong>2) Ordinal Variables:</strong> Ordinal variables (<span class="math inline">\(r \ge 2\)</span> categories) are generated by discretizing the standard normal variables at quantiles. These quantiles are determined by evaluating the inverse standard normal cdf at the cumulative probabilities defined by each variable’s marginal distribution, as follows: Suppose the goal is to simulate <span class="math inline">\(m\)</span> ordinal random variables for sample size <span class="math inline">\(n\)</span>. For <span class="math inline">\(1 \leq i \leq m\)</span>, let <span class="math inline">\({P}_{i}(y)\)</span> be the probability mass function and <span class="math inline">\({F}_{i}(y)\)</span> the cumulative distribution function of variable <span class="math inline">\({Y}_{i}\)</span>, with support <span class="math inline">\({\mathbb{Y}}_{i} = \{{y}_{i(1)},\ {y}_{i(2)},\ ...,\ {y}_{i(r_{i})}\}\)</span>. Here, <span class="math inline">\(r_{i}\)</span> is the number of categories for variable <span class="math inline">\({Y}_{i}\)</span>. Note that these are the <em>ordered</em> possible values for <span class="math inline">\({Y}_{i}\)</span> with associated probabilities: <span class="math display">\[{\mathbb{P}}_{i} = \{{P}_{i}({y}_{i(1)}),\ {P}_{i}({y}_{i(2)}),\ ...,\ {P}_{i}({y}_{i(r_{i})})\},\]</span> where <span class="math inline">\({P}_{i}({y}_{i(r_{i})}) = 1-\sum_{h=1}^{r_{i}-1}{P}_{i}({y}_{i(h)})\)</span>, and cumulative probabilities: <span class="math display">\[{\mathbb{F}}_{i} = \{{F}_{i}({y}_{i(l)}) = \sum_{j=1}^{l}{P}_{i}({y}_{i(j)}),\ l = 1,\ ...,\ r_{i}-1\}.\]</span> The final cumulative probability is obviously <span class="math inline">\(1\)</span>. The corresponding standard normal quantiles for these are given by: <span class="math display">\[{\mathbb{q}}_{i} = \{{q}_{i(l)} = {\Phi}^{-1}({F}_{i}({y}_{i(l)})) = {\Phi}^{-1}(\sum_{j=1}^{l}{P}_{i}({y}_{i(j)})),\ l = 1,\ ...,\ r_{i}-1\}.\]</span> If the support for variable <span class="math inline">\({Y}_{i}\)</span> is not provided, the default is to use <span class="math inline">\(\{1,\ 2,\ ...,\ r_{i}\}\)</span>. The <span class="math inline">\(m\)</span> random normal variates <span class="math inline">\({Z}_{1},\ {Z}_{2},\ ...,\ {Z}_{m}\)</span> are discretized using the cumulative probabilities and corresponding normal quantiles found above: <span class="math display">\[
\begin{aligned}
&amp;\textrm{if}\ {Z}_{i} &lt; {q}_{i(1)} \rightarrow {Y}_{i} = {y}_{i(1)}, \\
&amp;\textrm{if}\ {q}_{i(1)} \leq {Z}_{i} &lt; {q}_{i(2)} \rightarrow {Y}_{i} = {y}_{i(2)}, \\ 
&amp; \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ ..., \\
&amp;\textrm{if}\ {q}_{i(r_{i})} \leq {Z}_{i} \rightarrow {Y}_{i} = {y}_{i(r_{i})}.
\end{aligned}
\]</span></p>
<ol style="list-style-type: lower-alpha">
<li><p>The <strong>required parameters</strong> for simulating ordinal variables include: the cumulative marginal probabilities and support values (if desired). The probabilities should be combined into a list of length equal to the number of ordinal variables. The <span class="math inline">\(i^{th}\)</span> element is a vector of the cumulative probabilities defining the marginal distribution of the <span class="math inline">\(i^{th}\)</span> variable. If the variable can take <span class="math inline">\(r\)</span> values, the vector will contain <span class="math inline">\(r - 1\)</span> probabilities (the <span class="math inline">\(r^{th}\)</span> is assumed to be <span class="math inline">\(1\)</span>).</p></li>
<li><p>For <strong>binary variables</strong>, the user-supplied probability should be the probability of the <span class="math inline">\(1^{st}\)</span> (lower) support value. This would ordinarily be considered the probability of <em>failure</em> (<span class="math inline">\(q\)</span>), while the probability of the <span class="math inline">\(2^{nd}\)</span> (upper) support value would be considered the probability of <em>success</em> (<span class="math inline">\(p = 1 - q\)</span>). The support values should be combined into a separate list. The <span class="math inline">\(i^{th}\)</span> element is a vector containing the <span class="math inline">\(r\)</span> ordered support values. If not provided, the default is for the <span class="math inline">\(i^{th}\)</span> element to be the vector <span class="math inline">\(1, ..., r\)</span>.</p></li>
</ol>
<p><strong>3) Poisson and Negative Binomial Variables:</strong> Count variables are generated using the inverse cdf method. The cumulative distribution function of a standard normal variable has a uniform distribution. The appropriate quantile function <span class="math inline">\(F_{Y}^{-1}\)</span> is applied to this uniform variable with the designated parameters to generate the count variable: <span class="math inline">\(Y = F_{y}^{-1}(\Phi(Z))\)</span>.</p>
<p><strong>Zero-inflated distributions:</strong> A zero-inflated (ZI) random variable <span class="math inline">\(Y_{ZI}\)</span> is a mixture of a degenerate distribution having the point mass at <span class="math inline">\(0\)</span> and another distribution <span class="math inline">\(Y\)</span> that contributes both <span class="math inline">\(0\)</span> and non-<span class="math inline">\(0\)</span> values. If the mixing probability is <span class="math inline">\(\phi\)</span>, then: <span class="math display">\[P(Y_{ZI} = 0) = \phi + (1 - \phi) * P(Y = 0),\ \ 0 &lt; \phi &lt; 1.\]</span> Therefore, <span class="math inline">\(\phi\)</span> is the probability of a structural zero, and setting <span class="math inline">\(\phi = 0\)</span> reduces <span class="math inline">\(Y_{ZI}\)</span> to the variable <span class="math inline">\(Y\)</span>. In <strong>SimCorrMix</strong>, <span class="math inline">\(Y\)</span> can have either a Poisson distribution (<span class="math inline">\(Y_P\)</span>) or a NB distribution (<span class="math inline">\(Y_{NB}\)</span>) <span class="citation">(Ismail and Zamani 2013; Lambert 1992; Zhang, Mallick, and Yi 2016)</span>.</p>
<ol style="list-style-type: lower-alpha">
<li><strong>Zero-inflated Poisson (ZIP):</strong> The model for <span class="math inline">\(Y_{ZIP} \sim ZIP(\lambda,\ \phi),\ \lambda&gt;0,\ 0 &lt; \phi &lt; 1\)</span> is:<br />
<span class="math display">\[
\begin{aligned}
P(Y_{ZIP} = 0) &amp;= \phi + (1 - \phi) * exp(-\lambda) \\
P(Y_{ZIP} = y) &amp;= (1 - \phi) * exp(-\lambda) * \frac{\lambda^y}{y!},\ \ y = 1, 2, ...
\end{aligned}
\]</span> The mean of <span class="math inline">\(Y_{ZIP}\)</span> is <span class="math inline">\((1 - \phi) * \lambda\)</span>, and the variance is <span class="math inline">\(\lambda + \lambda^2 * \phi/(1 - \phi)\)</span> (see <code>VGAM::dzipois</code>).</li>
</ol>
<p>The zero-deflated Poisson distribution may be obtained by setting <span class="math inline">\(\phi \in (-1/(exp(\lambda) - 1),\ 0)\)</span>, so that the probability of a zero count is less than the nominal Poisson value. In this case, <span class="math inline">\(\phi\)</span> no longer represents a probability.</p>
<p>When <span class="math inline">\(\phi = -1/(exp(\lambda) - 1)\)</span>, the random variable has a positive-Poisson distribution (see <code>VGAM::dpospois</code>). The probability of a zero response is <span class="math inline">\(0\)</span>, and the other probabilities are scaled to sum to <span class="math inline">\(1\)</span>.</p>
<p>The parameters <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(\phi\)</span> are specified through the inputs <code>lam</code> and <code>p_zip</code>. Setting <code>p_zip = 0</code> (default setting) generates a regular Poisson variable. Parameters for zero-inflated Poisson variables should follow those for regular Poisson variables in all function inputs. For <strong>Correlation Method 2</strong>, a vector of total cumulative probability truncation values should be given in <code>pois_eps</code>. These values represent the amount removed from the total cumulative probability when creating finite supports. The value may differ by variable, but the default value is <span class="math inline">\(0.0001\)</span> (suggested by <span class="citation">Barbiero and Ferrari (2015)</span>). For example, <code>pois_eps = 0.0001</code> means that the support values removed have a total probability of <span class="math inline">\(0.0001\)</span> of occurring in the distribution of that variable. The effect is to remove improbable values, which may be of concern if the user wishes to replicate a distribution with outliers.</p>
<ol start="2" style="list-style-type: lower-alpha">
<li><strong>Zero-inflated Negative Binomial (ZINB):</strong> The model for <span class="math inline">\(Y_{ZINB} \sim ZINB(size,\ p,\ \phi),\ size&gt;0,\ 0&lt;p\leq1,\ \ 0 &lt; \phi &lt; 1\)</span> is:<br />
<span class="math display">\[
\begin{aligned}
P(Y_{ZINB} = 0) &amp;= \phi + (1 - \phi) * p^{size} \\
P(Y_{ZINB} = y) &amp;= (1 - \phi) * \frac{\Gamma(y + size)}{\Gamma(size) * y!} * p^{size} * (1 - p)^{size},\ \ y = 1, 2, ...
\end{aligned}
\]</span> In this formulation, the Negative Binomial component <span class="math inline">\(Y_{NB}\)</span> represents the number of failures which occur in a sequence of independent Bernoulli trials before a target number of successes (<span class="math inline">\(size\)</span>) is reached. The probability of success in each trial is <span class="math inline">\(p\)</span>. <span class="math inline">\(Y_{NB}\)</span> may also be parametrized by the mean <span class="math inline">\(\mu\)</span> and dispersion parameter <span class="math inline">\(size\)</span> so that <span class="math inline">\(p = size/(size + \mu)\)</span> or <span class="math inline">\(\mu = size * (1-p)/p\)</span> (see <code>stats::dnbinom</code>).</li>
</ol>
<p>The mean of <span class="math inline">\(Y_{ZINB}\)</span> is <span class="math inline">\((1 - \phi) * \mu\)</span>, and the variance is <span class="math inline">\((1 - \phi) * \mu * (1 + \mu * (\phi + 1/size))\)</span> (see <code>VGAM::dzinegbin</code>).<br />
The zero-deflated NB distribution may be obtained by setting <span class="math inline">\(\phi \in (-p^{size}/(1 - p^{size}),\ 0)\)</span>, so that the probability of a zero count is less than the nominal NB value. Again, in this case, <span class="math inline">\(\phi\)</span> no longer represents a probability.</p>
<p>The positive-NB distribution can be obtained with <span class="math inline">\(\phi = -p^{size}/(1 - p^{size})\)</span> (see <code>VGAM::dposnegbin</code>). The probability of a zero response is <span class="math inline">\(0\)</span>, and the other probabilities are scaled to sum to <span class="math inline">\(1\)</span>.</p>
<p>The parameters <span class="math inline">\(size\)</span>, <span class="math inline">\(p\)</span>, <span class="math inline">\(\mu\)</span>, and <span class="math inline">\(\phi\)</span> are specified through the inputs <code>size</code>, <code>prob</code>, <code>mu</code>, and <code>p_zinb</code>. Either <code>prob</code> or <code>mu</code> should be given for all NB and ZINB variables. Setting <code>p_zinb = 0</code> (default setting) generates a regular NB variable. Parameters for zero-inflated NB variables should follow those for regular NB variables in all function inputs. For <strong>Correlation Method 2</strong>, a vector of total cumulative probability truncation values should be given in <code>nb_eps</code>. The default value in the simulation functions is <span class="math inline">\(0.0001\)</span>.</p>
<p>The distributions functions are taken from the <strong>VGAM</strong> package <span class="citation">(Yee 2017)</span>.</p>
<div id="references" class="section level2">
<h2>References</h2>
<script type="text/x-mathjax-config">
   MathJax.Hub.Config({  "HTML-CSS": { minScaleAdjust: 115, availableFonts: [] }  });
</script>
<div id="refs" class="references">
<div id="ref-FerrBarb_Pois">
<p>Barbiero, A, and P A Ferrari. 2015. “Simulation of Correlated Poisson Variables.” <em>Applied Stochastic Models in Business and Industry</em> 31: 669–80. doi:<a href="https://doi.org/10.1002/asmb.2072">10.1002/asmb.2072</a>.</p>
</div>
<div id="ref-Dave">
<p>Davenport, JW, JC Bezder, and RJ Hathaway. 1988. “Parameter Estimation for Finite Mixture Distributions.” <em>Computers &amp; Mathematics with Applications</em> 15 (10): 819–28.</p>
</div>
<div id="ref-SMCD">
<p>Fialkowski, A C. 2017. <em>SimMultiCorrData: Simulation of Correlated Data with Multiple Variable Types</em>. <a href="https://CRAN.R-project.org/package=SimMultiCorrData" class="uri">https://CRAN.R-project.org/package=SimMultiCorrData</a>.</p>
</div>
<div id="ref-Fleish">
<p>Fleishman, A I. 1978. “A Method for Simulating Non-Normal Distributions.” <em>Psychometrika</em> 43: 521–32. doi:<a href="https://doi.org/10.1007/BF02293811">10.1007/BF02293811</a>.</p>
</div>
<div id="ref-Head2002">
<p>Headrick, T C. 2002. “Fast Fifth-Order Polynomial Transforms for Generating Univariate and Multivariate Non-Normal Distributions.” <em>Computational Statistics and Data Analysis</em> 40 (4): 685–711. doi:<a href="https://doi.org/10.1016/S0167-9473(02)00072-5">10.1016/S0167-9473(02)00072-5</a>.</p>
</div>
<div id="ref-HeadKow">
<p>Headrick, T C, and R K Kowalchuk. 2007. “The Power Method Transformation: Its Probability Density Function, Distribution Function, and Its Further Use for Fitting Data.” <em>Journal of Statistical Computation and Simulation</em> 77: 229–49. doi:<a href="https://doi.org/10.1080/10629360600605065">10.1080/10629360600605065</a>.</p>
</div>
<div id="ref-Ismail">
<p>Ismail, N, and H Zamani. 2013. “Estimation of Claim Count Data Using Negative Binomial, Generalized Poisson, Zero-Inflated Negative Binomial and Zero-Inflated Generalized Poisson Regression Models.” <em>Casualty Actuarial Society E-Forum</em> 41 (20): 1–28.</p>
</div>
<div id="ref-Lambert">
<p>Lambert, D. 1992. “Zero-Inflated Poisson Regression, with an Application to Defects in Manufacturing.” <em>Technometrics</em> 34 (1): 1–14.</p>
</div>
<div id="ref-Pears">
<p>Pearson, RK. 2011. “Exploring Data in Engineering, the Sciences, and Medicine.” In. New York: Oxford University Press.</p>
</div>
<div id="ref-Schork">
<p>Shork, N J, D B Allison, and B Thiel. 1996. “Mixture Distributions in Human Genetics Research.” <em>Statistical Methods in Medical Research</em> 5: 155–78. doi:<a href="https://doi.org/10.1177/096228029600500204">10.1177/096228029600500204</a>.</p>
</div>
<div id="ref-VGAM">
<p>Yee, T W. 2017. <em>VGAM: Vector Generalized Linear and Additive Models</em>. <a href="https://CRAN.R-project.org/package=VGAM" class="uri">https://CRAN.R-project.org/package=VGAM</a>.</p>
</div>
<div id="ref-Zhang">
<p>Zhang, X, H Mallick, and N Yi. 2016. “Zero-Inflated Negative Binomial Regression for Differential Abundance Testing in Microbiome Studies.” <em>Journal of Bioinformatics and Genomics</em> 2 (2): 1–9. doi:<a href="https://doi.org/10.18454/jbg.2016.2.2.1">10.18454/jbg.2016.2.2.1</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
